### PR TITLE
 BWS-108: + backend endpoint for upload object to space

### PR DESCRIPTION
**Pull request by ania-leichenko**
Screenshot_64
Screenshot_65

### CODE REVIEW

File: backend/src/api/bs/bs.api.ts [hunk 1]
```diff
@@ -88,6 +103,35 @@ const initBsApi: FastifyPluginAsync<Options> = async (fastify, opts) => {
...
+      const user: TokenPayload = await tokenService.decode(token);
+      const { id } = req.params;
+
+      const space = await spaceService.getSpacesById(id);
+
+      const worker = await workerService.getUserById(space.createdBy);
+
+      if (worker?.user?.tenantId !== user.tenantId || !req.file) {
+       return rep.send(false).status(HttpCode.BAD_REQUEST);
+      }
```
Comment (Yaroslaww-1) [inline backend/src/api/bs/bs.api.ts]: This code should be located in service

File: backend/src/data/models/object-upload/object-upload-table-field.enum.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+enum ObjectUploadField {
```
Comment (Yaroslaww-1) [inline backend/src/data/models/object-upload/object-upload-table-field.enum.ts]:
```suggestion
enum BSObjectField {
```

File: backend/src/data/models/object-upload/object-upload.model.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { TableName } from '~/common/enums/enums';
+import { AbstractModel } from '../abstract/abstract.model';
+import { ObjectUploadField } from './object-upload-table-field.enum';
+
+class ObjectUpload extends AbstractModel {
```
Comment (Yaroslaww-1) [inline backend/src/data/models/object-upload/object-upload.model.ts]:
```suggestion
class BSObject extends AbstractModel {
```

File: backend/src/data/repositories/object-upload/object-upload.repository.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import {
+  ObjectUpload as ObjectUploadM,
+} from '~/data/models/models';
+import { ObjectUpload as ObjectUploadEntity } from '~/services/object-upload/object-upload.entity';
+
+type Constructor = {
+  ObjectUploadModel: typeof ObjectUploadM;
+};
+
+class ObjectUpload {
```
Comment (Yaroslaww-1) [inline backend/src/data/repositories/object-upload/object-upload.repository.ts]:
```suggestion
class BSObject {
```

File: backend/src/data/models/object-upload/object-upload-table-field.enum.ts (hunk 2 above)
Comment (Yaroslaww-1) [reply backend/src/data/models/object-upload/object-upload-table-field.enum.ts]: Please rename `ObjectUpload` to `BSObject` everywhere

File: backend/src/middlewares/upload.middleware.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import multer from 'fastify-multer';
+
+const storage = multer.memoryStorage();
+
+const upload = multer({
+  storage,
+  limits: {
+    fileSize: 10000000,
```
Comment (Yaroslaww-1) [inline backend/src/middlewares/upload.middleware.ts]: It's better to set constants like that in this way
```suggestion
    fileSize: 10 * 1024 * 1024, // 10 MB
```

File: backend/src/services/aws/s3/s3.service.ts [hunk 6]
```diff
@@ -31,6 +33,18 @@ class S3 {
...
+    key,
+  }: {
+    name: string;
+    file: Buffer;
+    key: string;
+  }): Promise<PutObjectCommandOutput> {
+    return this.#s3Client.send(
+      new PutObjectCommand({ Bucket: name, Body: file, Key: key }),
+    );
+  }
```
Comment (Yaroslaww-1) [inline backend/src/services/aws/s3/s3.service.ts]: You already have BSObject service. What's the purpose of this method? If you need s3Client in BSObject service just inject it there

File: backend/src/services/object-upload/object-upload.service.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    user,
+  }: UploadPayload): Promise<ObjectUploadM> {
+    const { ETag: awsObjectKey } = await this.#s3Service.uploadObject({
+      name: space.name,
+      file: file.buffer,
+      key: file.originalname,
+    });
+
+    if (!awsObjectKey) {
+      throw Error('Error while uplosding file');
```
Comment (Yaroslaww-1) [inline backend/src/services/object-upload/object-upload.service.ts]:
```suggestion
      throw new BSCError({
        status: HttpCode.BAD_REQUEST,
        message: ExceptionMessage.OBJECT_NOT_UPLOADED,
      });
```

File: frontend/src/components/bs/bs.tsx [hunk 8]
```diff
@@ -8,6 +8,21 @@ import { bs as BSActions } from 'store/actions';
...
+    }
+    const spaceId = '5148d1ab-3929-4403-8922-f5d86cddbc11';
+    fetch(`http://localhost:3001/api/v1/bs/spaces/${spaceId}/objects`, {
+      method: 'POST',
+      headers: {
+        authorization: `Bearer ${localStorage.getItem('token')}`,
+      },
+      body: formData,
+    }).then((response) => response.json());
+  };
```
Comment (Yaroslaww-1) [inline frontend/src/components/bs/bs.tsx]: Move this code to service

**Review by Yaroslaww-1 — CHANGES_REQUESTED** [review]

File: frontend/src/components/bs/bs.tsx (hunk 8 above)
Comment (ania-leichenko) [reply frontend/src/components/bs/bs.tsx]: I left the front part for testing, when the edits are over I will remove it.

File: backend/src/services/aws/s3/s3.service.ts (hunk 6 above)
Comment (ania-leichenko) [reply backend/src/services/aws/s3/s3.service.ts]: this method encapsulates logic from s3client.

Comment (Donymak) [reply backend/src/services/aws/s3/s3.service.ts]: > You already have BSObject service. What's the purpose of this method? If you need s3Client in BSObject service just inject it there

Space Service uses S3 service to work with AWS, Why then BSObject service must do it in a different way?

Comment (Yaroslaww-1) [reply backend/src/services/aws/s3/s3.service.ts]: Oh, I missed that it is an aws service 😕

File: backend/src/data/repositories/bs-object/bs-object.repository.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+};
+
+class BSObject {
+  #BSObjectModel: typeof BSObjectM;
+
+  constructor({ BSObjectModel }: Constructor) {
+    this.#BSObjectModel = BSObjectModel;
+  }
+
+  async create(bsObject: BSObjectEntity): Promise<BSObjectM> {
```
Comment (Yaroslaww-1) [inline backend/src/data/repositories/bs-object/bs-object.repository.ts]:
```suggestion
  async create(bsObject: BSObjectEntity): Promise<BSObjectEntity> {
```

File: backend/src/services/bs-object/bs-object.service.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    workerService,
+  }: Constructor) {
+    this.#bsObjectRepository = bsObjectRepository;
+    this.#s3Service = s3Service;
+    this.#spaceService = spaceService;
+    this.#tokenService = tokenService;
+    this.#workerService = workerService;
+  }
+
+  public async upload({ token, file, id }: UploadPayload): Promise<BSObjectM> {
```
Comment (Yaroslaww-1) [inline backend/src/services/bs-object/bs-object.service.ts]:
```suggestion
  public async upload({ token, file, id }: UploadPayload): Promise<BSObjectEntity> {
```

**Review by Yaroslaww-1 — CHANGES_REQUESTED** [review]

File: backend/package.json [hunk 11]
```diff
@@ -30,6 +30,7 @@
     "dotenv": "12.0.4",
     "fastify": "3.25.3",
     "fastify-cors": "6.0.2",
+    "fastify-multer": "^2.0.2",
```
Comment (what1s1ove) [inline backend/package.json]: *B19*

File: backend/src/services/bs-object/bs-object.service.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    const user: TokenPayload = await this.#tokenService.decode(token);
+
+    const space = await this.#spaceService.getSpacesById(id);
+
+    const worker = await this.#workerService.getUserById(space.createdBy);
+
+    if (worker?.user?.tenantId !== user.tenantId || !file) {
```
Comment (what1s1ove) [inline backend/src/services/bs-object/bs-object.service.ts]: create variable for it

File: backend/src/services/bs-object/bs-object.service.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    this.#s3Service = s3Service;
+    this.#spaceService = spaceService;
+    this.#tokenService = tokenService;
+    this.#workerService = workerService;
+  }
+
+  public async upload({ token, file, id }: UploadPayload): Promise<BSObjectM> {
+    const user: TokenPayload = await this.#tokenService.decode(token);
+
+    const space = await this.#spaceService.getSpacesById(id);
```
Comment (what1s1ove) [inline backend/src/services/bs-object/bs-object.service.ts]: you should get user from request

File: backend/src/services/bs-object/bs-object.service.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+type Constructor = {
+  bsObjectRepository: typeof bsObjectRep;
+  s3Service: typeof s3Serv;
+  spaceService: typeof spaceServ;
+  tokenService: typeof tokenServ;
+  workerService: typeof workerServ;
+};
+
+type UploadPayload = {
```
Comment (what1s1ove) [inline backend/src/services/bs-object/bs-object.service.ts]: remove this type from this file

File: backend/src/api/api.ts [hunk 15]
```diff
@@ -63,6 +64,7 @@ const initApi: FastifyPluginAsync = async (fastify) => {
   fastify.register(initBsApi, {
     services: {
       space,
+      bsObject,
```
Comment (what1s1ove) [inline backend/src/api/api.ts]: remove bs prefix

File: backend/src/middlewares/bs-object.middleware.ts [hunk 16]
```diff
@@ -0,0 +1,12 @@
+import multer from 'fastify-multer';
+
+const storage = multer.memoryStorage();
+
+const upload = multer({
+  storage,
+  limits: {
+    fileSize: 10 * 1024 * 1024,
```
Comment (what1s1ove) [inline backend/src/middlewares/bs-object.middleware.ts]: add constant

File: backend/src/middlewares/bs-object.middleware.ts [hunk 17]
```diff
@@ -0,0 +1,12 @@
+import multer from 'fastify-multer';
```
Comment (what1s1ove) [inline backend/src/middlewares/bs-object.middleware.ts]: rename to `backend/src/middlewares/upload/upload.middleware.ts`

File: backend/src/middlewares/middlewares.ts [hunk 18]
```diff
@@ -0,0 +1 @@
+export * from './bs-object.middleware';
```
Comment (what1s1ove) [inline backend/src/middlewares/middlewares.ts]:
```suggestion
export {  upload } from './upload/upload.middleware';
```

File: backend/src/data/models/bs-object/bs-object.model.ts [hunk 19]
```diff
@@ -0,0 +1,18 @@
+import { TableName } from '~/common/enums/enums';
+import { AbstractModel } from '../abstract/abstract.model';
+import { BSObjectField } from './bs-object-table-field.enum';
+
+class BSObject extends AbstractModel {
```
Comment (what1s1ove) [inline backend/src/data/models/bs-object/bs-object.model.ts]: remove bs prefix

File: backend/src/data/models/bs-object/bs-object-table-field.enum.ts [hunk 20]
```diff
@@ -0,0 +1,10 @@
+enum BSObjectField {
```
Comment (what1s1ove) [inline backend/src/data/models/bs-object/bs-object-table-field.enum.ts]: remove bs prefix

File: backend/src/data/repositories/bs-object/bs-object.repository.ts [hunk 21]
```diff
@@ -0,0 +1,31 @@
+import { BSObject as BSObjectM } from '~/data/models/models';
+import { BSObject as BSObjectEntity } from '~/services/bs-object/bs-object.entity';
+
+type Constructor = {
+  BSObjectModel: typeof BSObjectM;
+};
+
+class BSObject {
```
Comment (what1s1ove) [inline backend/src/data/repositories/bs-object/bs-object.repository.ts]: remove bs prefix

File: backend/src/services/bs-object/bs-object.service.ts [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+  token: string;
+  file: {
+    buffer: Buffer;
+    originalname: string;
+    size: number;
+  };
+  id: string;
+};
+class BSObject {
```
Comment (what1s1ove) [inline backend/src/services/bs-object/bs-object.service.ts]: remove bs prefix

File: backend/src/api/bs/bs.api.ts [hunk 23]
```diff
@@ -88,6 +93,26 @@ const initBsApi: FastifyPluginAsync<Options> = async (fastify, opts) => {
       return rep.send(true).status(HttpCode.OK);
     },
   });
+
+  fastify.route({
+    method: HttpMethod.POST,
+    url: BSApiPath.SPACES_FILE,
+    preHandler: upload.single('file'),
+    async handler(
+      req: any, //eslint-disable-line
```
Comment (what1s1ove) [inline backend/src/api/bs/bs.api.ts]: import { File } from 'fastify-multer/lib/interfaces'; you can take this type from here

File: backend/src/api/bs/bs.api.ts [hunk 24]
```diff
@@ -88,6 +93,26 @@ const initBsApi: FastifyPluginAsync<Options> = async (fastify, opts) => {
...
+      rep,
+    ) {
+      const [, token] = req.headers?.authorization?.split(' ') ?? [];
+      const { id } = req.params;
+
+      await bsObjectService.upload({
+        token,
+        file: req.file,
```
Comment (what1s1ove) [inline backend/src/api/bs/bs.api.ts]: you need to add this property globally to FastifyRequest
https://github.com/fastify/fastify/discussions/3247

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/api/api.ts (hunk 15 above)
Comment (Yaroslaww-1) [reply backend/src/api/api.ts]: object is not a valid name for a class

Comment (Yaroslaww-1) [reply backend/src/api/api.ts]: the same story as with slc-function

Comment (what1s1ove) [reply backend/src/api/api.ts]: ahh got it, thanks

File: backend/src/services/bs-object/bs-object.service.ts (hunk 13 above)
Comment (ania-leichenko) [reply backend/src/services/bs-object/bs-object.service.ts]: why do we need to take user from request if everyone who did before got it from token ?
https://github.com/BinaryStudioAcademy/bsa-winter-2021-2022-bws/blob/development/backend/src/services/worker/worker.service.ts#L193

File: backend/src/common/types/bs-object/bs-object.ts [hunk 25]
```diff
@@ -0,0 +1,9 @@
+import { File } from 'fastify-multer/lib/interfaces';
```
Comment (what1s1ove) [inline backend/src/common/types/bs-object/bs-object.ts]: move to `backend/src/common/types/bs-object/bs-object-upload-payload.type.ts`

File: backend/src/plugins/bs-object/bs-object.plugin.ts [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+  interface FastifyRequest {
+    file: File;
+  }
+}
+
+export interface FilePluginOptions {
+  file: File;
+}
+
+const filePlugin: FastifyPluginAsync<FilePluginOptions> = fp(
```
Comment (what1s1ove) [inline backend/src/plugins/bs-object/bs-object.plugin.ts]:
```suggestion
const file: FastifyPluginAsync<FilePluginOptions> = fp(
```

Comment (what1s1ove) [inline backend/src/plugins/bs-object/bs-object.plugin.ts]:
```suggestion
const file: FastifyPluginAsync<{
  file: File;
}> = fp(
```

File: backend/src/plugins/bs-object/bs-object.plugin.ts [hunk 27]
```diff
@@ -0,0 +1,21 @@
+import { FastifyPluginAsync } from 'fastify';
+import fp from 'fastify-plugin';
+import { File } from 'fastify-multer/lib/interfaces';
+
+declare module 'fastify' {
```
Comment (what1s1ove) [inline backend/src/plugins/bs-object/bs-object.plugin.ts]: you should to declare it in the `src/index.d.ts`

File: backend/src/services/bs-object/bs-object.service.ts [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+    id,
+  }: UploadPayload): Promise<BSObjectEntity> {
+    const user: TokenPayload = await this.#tokenService.decode(token);
+
+    const space = await this.#spaceService.getSpacesById(id);
+
+    const worker = await this.#workerService.getUserById(space.createdBy);
+    const notCurrentWorker = worker?.user?.tenantId !== user.tenantId;
+
+    if (notCurrentWorker || !file) {
```
Comment (what1s1ove) [inline backend/src/services/bs-object/bs-object.service.ts]:
```suggestion
    const isCurrentWorker = worker?.user?.tenantId === user.tenantId;

    if (!isCurrentWorker || !hasFile) {
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]
- Root `package-lock.json` shouldn't be changed
- you need to remove all frontend stuff before merge

**Pull request closed without merging**